Added docs for workflow versioning (#5003)
dapr/docs: daprdocs/content/en/developing-applications/building-blocks/workflow/workflow-versioning.md

In either versioning strategy, workflows can reach the `stalled` state. This typically occurs during rollouts, when both old and new versions of the workflow code are running at the same time.

A versioned workflow will become stalled when it is started on a new replica, but a subsequent replay is attempted
on an old replica.

Workflows can remain stalled for the entire duration of the rollout, until there are only new replicas available. Once a stalled workflow is scheduled on a new replica, it will continue executing.

If a workflow never continues it means the conditions for getting stalled are still present and will need to be
addressed. See the specific reasons for getting stalled in the each of the versioning approaches below.
{{% /alert %}}

### What Does Versioning Solve?
To best understand how versioning works and how to use it effectively, it helps to first understand what it's
solving and why versioning is necesssary at all. Workflows are required to be deterministic meaning that every time they are run, they produce precisely the same output as the last time.

This is a key concept because Dapr Workflows are implemented to use an event sourcing approach to persist the

> workflow state. As activities and child workflows are executed within the workflow, we're maintaining a history that

Reviewer: ```suggestion
workflow state. As activities and child workflows are executed within the workflow, Dapr is maintaining a history that
```

> logs the inputs and outputs at each of those boundary executions. As each completes, we persist the updated event

Reviewer: ```suggestion
logs the inputs and outputs at each of those boundary executions. As each completes, Dapr Workflow persist the updated event
```

history then re-invoke the workflow from the top (this is called a "replay"). This time, when it hits one of these
activities or child workflows, it substitutes in the already-realized output and skips re-executing it and repeats.

> Because of this, it's critical that no code be arbitrarily introduced into a workflow while it's in-flight because when the engine replays the workflow, your changes risk the workflow no longer matching with its history and failing.

Reviewer: ```suggestion
Because of this, it's critical that no code be arbitrarily introduced into a workflow while it's in-flight because when Dapr replays the workflow, your changes risk the workflow no longer matching with its history and failing.
```

        {
            // Patched code
            await context.CallActivityAsync(nameof(SendSMS), input);
        }
        else {
            // Original code
            await context.CallActivityAsync(nameof(SendEmail), input);
        }
        return null;
    }
}
```

{{% /tab %}}

{{% tab "Go" %}}

```go
import "github.com/dapr/durabletask-go/workflow"

func Workflow(ctx *workflow.WorkflowContext) error {
  // ...
  if ctx.IsPatched("use-sms") {
    if err := ctx.CallActivity("SendSMS", ctx.GetInput()).Await(); err != nil {
      return err
    }
  } else {
    if err := ctx.CallActivity("SendEmail", ctx.GetInput()).Await(); err != nil {
      return err
    }
  }
  // ...
}

```

{{% /tab %}}

{{% tab "Python" %}}

```python
from dapr.ext.workflow import DaprWorkflowContext, WorkflowRuntime

wfr = WorkflowRuntime()

@wfr.workflow
def my_workflow(ctx: DaprWorkflowContext, wf_input: str):
  # ...
  if ctx.is_patched("use-sms"):
    yield ctx.call_activity(send_sms)
  else:
    yield ctx.call_activity(send_email)
  # ...
```

{{% /tab %}}

{{< /tabpane >}}

Using this approach, new workflow instances will take the patched code path but existing workflows that are

> in-flight when the patch is introduced will continue to use the original code path. Patch checks are recorded in the

Reviewer: ```suggestion
in-flight when the patch is introduced continue to use the original code path. Patch checks are recorded in the
```

workflow instance history the first time they are evaluated, meaning that it's safe to use the same identifier in
different places throughout your workflow.

> Again, it's imperative that once your patched workflow is running in production, you do not repurpose that

Reviewer: ```suggestion
It's important that once your patched workflow is running in production, you do not repurpose that
```

> identifier again though. Once deployed, it should be assumed that the engine will handle the workflow

Reviewer: ```suggestion
identifier again though. Once deployed, it should be assumed that the Dapr handles the workflow
```

  workflows to evaluate in the same way.
- Patches should be applied in an additive way meaning that once they're added and your application deployed,
  you should never move nor remove them from your workflow. They must exist to retain in-flight workflow determinism.
- As covered above, you must not re-use a patch identifier that you've used in previous deployments as this will
  break the determinism guarantee. However, you're welcome to use that same identifier within the same deployment in
  multiple places and without concern across different workflows.
- If you can avoid using an `else` branch when applying your patch, it will make it easier to apply future patches.
  While this generally can't be avoided if you're replacing existing code, it's certainly manageable if you're
  simply adding new logic to your workflow.
- If you must nest patches, you must do so within existing patches. For example your new patch cannot wrap an
  existing patch else in the `else` branch. Doing so would other mean that the originally code cannot be accessed by
  in-flight workflows, breaking the determinism guarantee.

### Named Workflow Versioning

> Named workflow versioning represents our second approach to facilitating a means of versioning your workflows in a

Reviewer: ```suggestion
Named workflow versioning represents a second approach to versioning your workflows in a
```

deterministically safe way. In this approach, you introduce a new workflow version by duplicating an existing
workflow and assigning it a new "versioned" name. Because you can then refactor your workflow logic from scratch to
remove any patches and otherwise introduce any other changes you want, this approach offers a clean break from
previous workflows versions.

> While the particulars of how this is achieved will vary depending on the language SDK you're using, generally, the

Reviewer: ```suggestion
While the particulars of how this is achieved varies depending on the language SDK you're using, generally, the
```

idea is that you'll duplicate your latest workflow, assign it a unique name that reflects the newer version and

> register this with your SDK. A registry will be built of each workflow so that at runtime, Dapr will call in with a

Reviewer: ```suggestion
register this with your SDK. A registry is then built of each workflow so that at runtime, Dapr calls with a
```

> workflow name to execute and the SDK will either route to the old workflow (if in-flight) or to the newer version.

Reviewer: ```suggestion
workflow name to execute and the SDK either routes to the old workflow (if in-flight) or to the newer version.
```

> Do note that across all SDKs, the runtime does not migrate workflows between versions sequentially. Rather, when the

Reviewer: ```suggestion
Do note that across all SDKs, the Dapr runtime does not migrate workflows between versions sequentially. Rather, when the
```

> SDKs receive a request to run a new workflow, it will opt to run the latest version, not simply the "next" one, so

Reviewer: ```suggestion
SDKs receive a request to run a new workflow, it chooses to run the latest version, not simply the "next" one, so
```

there's is no need to handle any compensation logic between versions.

> The language SDKs may expose a way to register versions when using the same workflow name, but this will vary by SDK,

Reviewer: ```suggestion
The language SDKs may expose a way to register versions when using the same workflow name, but this varies by SDK,
```

> so please refer to the documentation for your specific SDK for more information. For example, some may use a mechanism

Reviewer: ```suggestion
so refer to the documentation for a specific SDK for more information. For example, some may use a mechanism
```

by which you would explicitly supply an `is_latest` flag to indicate which version is the latest.

When the SDK receives a request to run a workflow in a version that is not registered, the workflow will stall. This
can happen naturally during application rollouts, but can also happen if a version is removed while some workflow

> instances are still using it. Our recommendation is that older named workflow versions should not be deleted at all

Reviewer: ```suggestion
instances are still using it. The recommendation is that older named workflow versions should not be deleted at all
```
